core/src/main/java/com/scalar/db/transaction/consensuscommit/MutationsGrouper.java

```diff
@@ -0,0 +1,155 @@
+package com.scalar.db.transaction.consensuscommit;
+
+import com.scalar.db.api.Mutation;
+import com.scalar.db.api.StorageInfo;
+import com.scalar.db.common.StorageInfoProvider;
+import com.scalar.db.exception.storage.ExecutionException;
+import com.scalar.db.io.Key;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+public class MutationsGrouper {
+
+  private final StorageInfoProvider storageInfoProvider;
+
+  public MutationsGrouper(StorageInfoProvider storageInfoProvider) {
+    this.storageInfoProvider = storageInfoProvider;
+  }
+
+  public List<List<Mutation>> groupMutations(Collection<Mutation> mutations)
+      throws ExecutionException {
+    // MutationGroup mutations by their storage info and atomicity unit
+    List<MutationGroup> groups = new ArrayList<>();
+    Map<MutationGroup, List<List<Mutation>>> groupToBatches = new LinkedHashMap<>();
+
+    for (Mutation mutation : mutations) {
+      assert mutation.forNamespace().isPresent();
+      StorageInfo storageInfo = storageInfoProvider.getStorageInfo(mutation.forNamespace().get());
+
+      MutationGroup matchedGroup = null;
```

Comment:
[minor] If `MutationGroup.equals()` has a similar logic to `isSameGroup()`, maybe these lines can be simplified like this?

```java
      MutationGroup group = new MutationGroup(mutation, storageInfo);

      List<List<Mutation>> batches =
          groupToBatches.computeIfAbsent(group, g -> new ArrayList<>());
      int maxCount = group.storageInfo.getMaxAtomicMutationsCount();
```